src/components/breadcrumb/Breadcrumb.tsx

```diff
@@ -0,0 +1,16 @@
+import React from 'react'
+import { Switch, Route } from 'react-router'
+import DefaultBreadcrumb from 'components/breadcrumb/DefaultBreadcrumb'
+import PatientBreadcrumb from 'components/breadcrumb/PatientBreadcrumb'
+import AppointmentBreadcrumb from 'components/breadcrumb/AppointmentBreadcrumb'
+
+const Breadcrumb = () => (
+  <Switch>
+    <Route exact path={['/patients/new', '/appointments/new']} component={DefaultBreadcrumb} />
+    <Route path="/patients/:id" component={PatientBreadcrumb} />
+    <Route path="/appointments/:id" component={AppointmentBreadcrumb} />
+    <Route path="*" component={DefaultBreadcrumb} />
+  </Switch>
+)
```

Comment:
I'm wondering if we can combine all of the breadcrumbs into one component. 

This breadcrumb component would read from a Provider (defined using the React Context API) or a Redux Slice. 

The state the Provider or Slice would be an array of objects that have two properties: text and  location. 

The Provider or Slice would have three functions exposed, one to add a breadcrumb, one to remove a breadcrumb, and one to set the breadcrumbs (basically a function to just set the state exactly to what the user passed in). 

Each Route Handler (e.g. `Patients`, `ViewPatient`, `NewPatient`) would then add a breadcrumb to the state, and would remove it during the cleanup phase of the `useEffect`.

This would ensure we don't have any duplicate code, the breadcrumb can be a single component, and we don't have any additional work each time a new route or module is added.

Feel free to ping me on slack with any questions.

Reply:
For an example of how this might work, checkout https://github.com/HospitalRun/hospitalrun-frontend/pull/1817/files. I'm working through doing something similar but with a button toolbar.

Reply:
Ok, if my understanding is correct, each Route Component will declare an `effect` to update the breadcrumb array in the state ?

Reply:
I think that you have the correct understanding!

Reply:
I updated my code using your recommendations (without doing the unit tests).
I only handled the `/` and `/patients/**` routes, let met know if it's ok for you, then I will finish the tests and handle the `/appointments/**` routes

Reply:
I just finished to implement the whole feature (I still need to improve the code coverage). 

It was not as easy as I expected because of the nested routes. I realized the effect of a child component is executed before the effect of its parent. As a result I couldn't rely on the order of the addBreadcrumb function calls to build a valid breadcrumbs array. So I decided to sort the breadcrumbs by their location property before the rendering. 

I am also afraid that maintaining the breadcrumbs will be painful in the future, as the number of routes will increase. Each developers should remember that when they add a route, they should also call the hook `useAddBreadcrumbs`, it can be really cumbersome...